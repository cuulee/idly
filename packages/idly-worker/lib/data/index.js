export { wikipedia as dataWikipedia } from 'wmf-sitematrix';
export { default as dataSuggestions } from 'name-suggestion-index/name-suggestions.json';
export { dataAddressFormats } from 'data/address-formats.json';
export { dataDeprecated } from 'data/deprecated.json';
export { dataDiscarded } from 'data/discarded.json';
export { dataLocales } from 'data/locales.json';
export { dataPhoneFormats } from 'data/phone-formats.json';
export { dataShortcuts } from 'data/shortcuts.json';
export { default as dataImperial } from 'data/imperial.json';
export { default as dataDriveLeft } from 'data/drive-left.json';
// export { en as dataEn } from '../dist/locales/en.json';
import { dataImagery } from 'data/imagery.json';
import { categories } from 'data/presets/categories.json';
import { defaults } from 'data/presets/defaults.json';
import { fields } from 'data/presets/fields.json';
import { presets } from 'data/presets/presets.json';
// import maki from '@mapbox/maki';
// export var dataFeatureIcons = maki.layouts.all.all;
export let data = {
    imagery: dataImagery,
    presets: {
        presets,
        defaults,
        categories,
        fields
    }
};
//# sourceMappingURL=index.js.map