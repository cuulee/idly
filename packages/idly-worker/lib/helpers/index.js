export function isClosed(entity) {
    return (entity.nodes.length > 1 &&
        entity.nodes[0] === entity.nodes[entity.nodes.length - 1]);
}
//# sourceMappingURL=index.js.map