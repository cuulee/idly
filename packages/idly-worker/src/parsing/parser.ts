import { deepFreeze } from 'idly-common/lib/misc/deepFreeze';
import { ImMap, ImSet } from 'idly-common/lib/misc/immutable';
import { attributesGen } from 'idly-common/lib/osm/attributesGen';
import { genLngLat } from 'idly-common/lib/osm/genLngLat';
import { nodeFactory } from 'idly-common/lib/osm/nodeFactory';
import { relationFactory } from 'idly-common/lib/osm/relationFactory';
import { relationMemberGen } from 'idly-common/lib/osm/relationMemberGen';
import {
  Entity,
  EntityId,
  EntityTable,
  Node,
  ParentWays,
  Relation,
  Tags,
  Way
} from 'idly-common/lib/osm/structures';
import { tagsFactory } from 'idly-common/lib/osm/tagsFactory';
import { wayFactory } from 'idly-common/lib/osm/wayFactory';

export function calculateParentWays(parentWays: ParentWays, ways: Way[]) {
  console.time('Imm: calculateParentWays');
  const x = parentWays.withMutations(p => {
    ways.forEach(w => {
      w.nodes.forEach(nodeId => {
        p.update(nodeId, (s = ImSet()) => s.add(w.id));
      });
    });
  });
  console.timeEnd('Imm: calculateParentWays');
  return x;
}

const xmlIndex = process.env.NODE_ENV === 'test' ? 0 : 2;
/**
 * @NOTE parseXML doesnt care if there are duplicates
 *  as this helps in removing a particular tile
 *  and still maintaining the duplicated entity in other
 *  tile. If in case we started screening duplicates
 *  this might remove an entity which might span across
 *  multiple tiles.
 * @param xml
 * @param parentWays
 */
export function parseXML(
  xml: Document | undefined,
  parentWays: ParentWays = ImMap()
):
  | undefined
  | {
      entities: Entity[];
      parentWays: ParentWays;
    } {
  if (!xml || !xml.childNodes) return;
  const root = xml.childNodes[xmlIndex];

  const children = root.childNodes;
  const entities: Entity[] = [];
  const nodesXML: any = [];
  const waysXML: any = [];
  const relationsXML: any = [];
  const group: {
    node: Node[];
    way: Way[];
    relation: Relation[];
  } = {
    node: [],
    way: [],
    relation: []
  };
  for (let i = 0, l = children.length; i < l; i++) {
    const child = children[i];
    const parser = group[child.nodeName];
    if (parser) group[child.nodeName].push(child);
  }

  group.relation = group.relation.map(parsers.relation);
  group.way = group.way.map(w => parsers.way(w));

  // calculateParentWays right at this point allows us
  // to pass the parsed group.way which is needed to
  // calculate parentWays.
  // @NOTE calculateParentWays directly mutates
  //  I want to change this behaviour
  // parentWays = calculateParentWays(parentWays, group.way);

  parentWays = calculateParentWays(parentWays, group.way);

  group.node = group.node.map(n => parsers.node(n));

  return {
    entities: [...group.node, ...group.way, ...group.relation],
    parentWays
  };
}

function getVisible(attrs) {
  return (
    !attrs.getNamedItem('visible') ||
    attrs.getNamedItem('visible').value !== 'false'
  );
}

function getMembers(obj) {
  const elems = obj.getElementsByTagName('member');
  const members = new Array(elems.length);
  for (let i = 0, l = elems.length; i < l; i++) {
    const attrs = elems[i].attributes;
    members[i] = relationMemberGen({
      id: attrs.getNamedItem('type').value[0] + attrs.getNamedItem('ref').value,
      type: attrs.getNamedItem('type').value,
      role: attrs.getNamedItem('role').value
    });
  }
  return members;
}

function getNodes(obj) {
  const elems = obj.getElementsByTagName('nd');
  const nodes = new Array(elems.length);
  for (let i = 0, l = elems.length; i < l; i++) {
    nodes[i] = 'n' + elems[i].attributes.getNamedItem('ref').value;
  }
  return nodes;
}

function getTags(obj): Tags {
  const elems = obj.getElementsByTagName('tag');
  const t: Array<[string, string]> = [];
  for (let i = 0, l = elems.length; i < l; i++) {
    const attrs = elems[i].attributes;
    t.push([attrs.getNamedItem('k').value, attrs.getNamedItem('v').value]);
  }
  return tagsFactory(t);
}

function getLoc(attrs) {
  const lon = attrs.getNamedItem('lon') && attrs.getNamedItem('lon').value;
  const lat = attrs.getNamedItem('lat') && attrs.getNamedItem('lat').value;
  return genLngLat([parseFloat(lon), parseFloat(lat)]);
}

const parsers = {
  node: function nodeData(obj) {
    const attrs = obj.attributes;
    const id = 'n' + attrs.getNamedItem('id').value;
    return nodeFactory({
      id,
      attributes: attributesGen({
        visible: getVisible(attrs),
        version: attrs.getNamedItem('version').value,
        timestamp:
          attrs.getNamedItem('timestamp') &&
          attrs.getNamedItem('timestamp').value,
        changeset:
          attrs.getNamedItem('changeset') &&
          attrs.getNamedItem('changeset').value,
        uid: attrs.getNamedItem('uid') && attrs.getNamedItem('uid').value,
        user: attrs.getNamedItem('user') && attrs.getNamedItem('user').value
      }),
      loc: getLoc(attrs),
      tags: getTags(obj)
    });
  },

  way: function wayData(obj) {
    const attrs = obj.attributes;
    return wayFactory({
      id: 'w' + attrs.getNamedItem('id').value,
      attributes: attributesGen({
        visible: getVisible(attrs),
        version: attrs.getNamedItem('version').value,
        changeset:
          attrs.getNamedItem('changeset') &&
          attrs.getNamedItem('changeset').value,
        timestamp:
          attrs.getNamedItem('timestamp') &&
          attrs.getNamedItem('timestamp').value,
        user: attrs.getNamedItem('user') && attrs.getNamedItem('user').value,
        uid: attrs.getNamedItem('uid') && attrs.getNamedItem('uid').value
      }),
      tags: getTags(obj),
      nodes: getNodes(obj)
    });
  },

  relation: function relationData(obj) {
    const attrs = obj.attributes;
    return relationFactory({
      id: 'r' + attrs.getNamedItem('id').value,
      attributes: attributesGen({
        visible: getVisible(attrs),
        version: attrs.getNamedItem('version').value,
        changeset:
          attrs.getNamedItem('changeset') &&
          attrs.getNamedItem('changeset').value,
        timestamp:
          attrs.getNamedItem('timestamp') &&
          attrs.getNamedItem('timestamp').value,
        user: attrs.getNamedItem('user') && attrs.getNamedItem('user').value,
        uid: attrs.getNamedItem('uid') && attrs.getNamedItem('uid').value
      }),
      tags: getTags(obj),
      members: getMembers(obj)
    });
  }
};
