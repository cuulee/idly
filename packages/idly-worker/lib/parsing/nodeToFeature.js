import { getNodeGeometry } from 'helpers/getNodeGeometry';
import { presetsMatcherCached } from 'presets/presets';
export const DEFAULT_NODE_ICON = 'circle';
export function nodeCombiner(node, parentWays) {
    return Object.assign({}, nodeToPoint(node), { properties: Object.assign({}, applyNodeMarkup(
        // presetsMatcher,
        getNodeGeometry(node.id, parentWays), node.tags), { id: node.id }) });
}
export const nodeToPoint = (node) => {
    return {
        id: node.id,
        type: 'Feature',
        geometry: {
            type: 'Point',
            coordinates: [node.loc.lon, node.loc.lat]
        },
        properties: {}
    };
};
export const applyNodeMarkup = (geometry, tags) => {
    const match = presetsMatcherCached(geometry)(tags);
    return {
        icon: (match && match.icon) || DEFAULT_NODE_ICON,
        name: tags.get('name'),
        geometry
    };
};
//# sourceMappingURL=nodeToFeature.js.map