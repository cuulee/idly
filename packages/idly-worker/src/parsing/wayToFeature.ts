// import { LineString, Polygon } from 'geojson';
// import * as turf from 'turf';

// import { List, Map } from 'immutable';
// import { Feature } from 'typings/geojson';

// import { isArea } from 'osm/entities/helpers/misc';
// import { Node } from 'osm/entities/node';
// import { presetsMatcherCached } from 'osm/presets/presets';

// import { Geometry } from 'osm/entities/constants';
// import { Way } from 'osm/entities/way';
// import { Graph } from 'osm/history/graph';
// import { tagClassesPrimary } from 'osm/styling/tagClasses';
// import { weakCache } from 'utils/weakCache';
import * as turfLineString from 'turf-linestring';
import * as turfPolygon from 'turf-polygon';

import { LineString, Polygon } from 'geojson';
import { isArea } from 'helpers/isArea';
import { weakCache } from 'helpers/weakCache';
import { Geometry } from 'structs/geometry';
import { Node } from 'structs/node';
import { Way } from 'structs/way';

import { tagClassesPrimary } from 'osm/tagsStyling';
import { presetsMatcherCached } from 'presets/presets';
import { Table } from 'structs/table';
import { Feature } from 'types/geojson';

interface IWayProperties {
  id: string;
  icon: string | undefined;
  name?: string;
  geometry: Geometry.LINE | Geometry.AREA;
  tagsClass: string;
  tagsClassType: string;
}
/**
 * @TOFIX use areaKeys from the core.
 */
export type WayFeature = Feature<LineString | Polygon, IWayProperties>;

const tagClassesPrimaryCache = weakCache(tagClassesPrimary);

/**
 *
 * @REVISIT this func also needs to handle modifiedGraph
 *  for now i removed it so i can test caching and anyway
 *  not yet implemented modified ways
 */

export function wayCombiner(way: Way, table: Table) {
  const geometry = isArea(way) ? Geometry.AREA : Geometry.LINE;
  const [tagsClass, tagsClassType] = tagClassesPrimaryCache(way.tags);
  const match = presetsMatcherCached(geometry)(way.tags);
  // const nodes = getCoordsFromGraph(graph.node, way.nodes);

  const nodes = getCoordsFromTable(table, way.nodes);

  const properties: IWayProperties = {
    id: way.id,
    name: way.tags.get('name') || way.tags.get('ref'),
    icon: match && match.icon,
    geometry,
    tagsClass,
    tagsClassType
  };
  return {
    id: way.id,
    ...wayToLineString(geometry, nodes),
    properties
  };
}
export function getCoordsFromTable(table: Table, nodes: string[]): number[][] {
  return nodes.map(n => {
    if (!table.has(n)) throw new Error('node not found ' + n);
    const node = table.get(n) as Node;
    return [node.loc.lon, node.loc.lat];
  });
}

/**
 * graphNode comes from graph.node
 *  planning to use it for memoization.
 *  every ms counts. hehehe.
 */
export function wayToLineString(geometry: Geometry, nodeCoords: number[][]) {
  let feat;
  if (geometry === Geometry.LINE) {
    feat = turfLineString(nodeCoords, {}) as Feature<
      LineString,
      IWayProperties
    >;
  } else if (geometry === Geometry.AREA) {
    feat = turfPolygon([nodeCoords], {}) as Feature<Polygon, IWayProperties>;
  } else {
    throw new Error('not a matching geometry provided for way');
  }
  return feat;
}
