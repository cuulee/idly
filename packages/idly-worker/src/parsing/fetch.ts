import * as SphericalMercator from '@mapbox/sphericalmercator';
import { handleErrors } from 'helpers/promise';
import { stubXML } from 'parsing/fixtures';
const DOMParser = require('xmldom').DOMParser;

const merc = new SphericalMercator({
  size: 256
});

const fetchStub = () => {
  return Promise.resolve(stubXML);
};
const debug = false;

export async function fetchTile(x: number, y: number, zoom: number) {
  const xyz = [x, y, zoom].join(',');
  const bboxStr = merc.bbox(x, y, zoom).join(',');
  try {
    if (debug) {
      return new DOMParser().parseFromString(await fetchStub(), 'text/xml');
    }
    let response = await fetch(
      `https://www.openstreetmap.org/api/0.6/map?bbox=${bboxStr}`
    );
    response = handleErrors(response);
    const text = await response.text();

    const parser = new DOMParser();
    const xml = parser.parseFromString(text, 'text/xml');
    return xml;
  } catch (e) {
    console.error(e);
    throw e;
  }
}
