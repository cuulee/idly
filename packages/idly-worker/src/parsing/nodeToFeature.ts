import { Feature, Point } from 'geojson';

import { getNodeGeometry } from 'helpers/getNodeGeometry';
import { ParentWays } from 'parsing/parser';
import { presetsMatcherCached } from 'presets/presets';
import { Geometry } from 'structs/geometry';
import { Node } from 'structs/node';
import { Tags } from 'structs/tags';

type NodeFeat = Feature<Point>;
type NodeGeometry = Geometry.POINT | Geometry.VERTEX | Geometry.VERTEX_SHARED;

interface INodeMarkup {
  icon: string;
  name?: string;
  geometry: NodeGeometry;
}
export const DEFAULT_NODE_ICON = 'circle';

export function nodeCombiner(node: Node, parentWays: ParentWays) {
  return {
    ...nodeToPoint(node),
    properties: {
      ...applyNodeMarkup(
        // presetsMatcher,
        getNodeGeometry(node.id, parentWays),
        node.tags
      ),
      id: node.id
    }
  };
}

export const nodeToPoint = (node: Node): Feature<Point> => {
  return {
    id: node.id,
    type: 'Feature',
    geometry: {
      type: 'Point',
      coordinates: [node.loc.lon, node.loc.lat]
    },
    properties: {}
  };
};

export const applyNodeMarkup = (geometry: NodeGeometry, tags: Tags) => {
  const match = presetsMatcherCached(geometry)(tags);
  return {
    icon: (match && match.icon) || DEFAULT_NODE_ICON,
    name: tags.get('name'),
    geometry
  };
};
