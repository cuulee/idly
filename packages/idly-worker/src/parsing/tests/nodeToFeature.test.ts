import { genLngLat, nodeFactory, tagsFactory } from 'idly-common/lib';

import { nodeCombiner } from '../../parsing/nodeToFeature';

const n1 = nodeFactory({ id: 'n-1' });
const n11 = nodeFactory({
  id: 'n-1',
  loc: genLngLat({ lon: 15, lat: 10 }),
  tags: tagsFactory({ k: 'k' })
});

describe('converts node to feat', () => {
  describe('nodeCombiner', () => {
    it('should behave...', () => {
      expect(nodeCombiner(n1, {})).toMatchSnapshot();
      expect(nodeCombiner(n11, {})).toMatchSnapshot();
    });
    it('should add computed Props', () => {
      expect(nodeCombiner(n1, { test: '1234', h: 3 })).toMatchSnapshot();
    });
  });
});
