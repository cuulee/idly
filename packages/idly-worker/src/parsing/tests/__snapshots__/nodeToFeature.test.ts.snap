// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`converts node to feat nodeCombiner should work 1`] = `
Object {
  "geometry": Object {
    "coordinates": Array [
      15,
      10,
    ],
    "type": "Point",
  },
  "id": "n-1",
  "properties": Object {
    "geometry": "point",
    "icon": "circle",
    "id": "n-1",
    "name": undefined,
  },
  "type": "Feature",
}
`;

exports[`converts node to feat nodeToPoint should behave... 1`] = `
Object {
  "geometry": Object {
    "coordinates": Array [
      0,
      0,
    ],
    "type": "Point",
  },
  "id": "n-1",
  "properties": Object {},
  "type": "Feature",
}
`;

exports[`converts node to feat nodeToPoint should behave... 2`] = `
Object {
  "geometry": Object {
    "coordinates": Array [
      15,
      10,
    ],
    "type": "Point",
  },
  "id": "n-1",
  "properties": Object {},
  "type": "Feature",
}
`;
